lib/pbench/server/api/config.py

```diff
@@ -0,0 +1,109 @@
+import os
+import sys
+from pathlib import Path
+from pbench.server import PbenchServerConfig
+from pbench.common.exceptions import BadConfig
+from pbench.server.utils import filesize_bytes
+from pbench.common.logger import get_pbench_logger
+
+
+def get_config(cfg_name):
+    if not cfg_name:
+        print(
+            f"{__name__}: ERROR: No config file specified; set"
+            " _PBENCH_SERVER_CONFIG",
+            file=sys.stderr,
+        )
+        sys.exit(1)
+
+    try:
+        return PbenchServerConfig(cfg_name)
+    except BadConfig as e:
+        print(f"{__name__}: {e} (config file {cfg_name})", file=sys.stderr)
+        sys.exit(1)
+
+
+def get_prdp(config_server, logger):
+    prdp = config_server.get("pbench-receive-dir-prefix")
+    if not prdp:
+        logger.error("Missing config variable for pbench-receive-dir-prefix")
+        sys.exit(1)
+    try:
+        upload_directory = Path(f"{prdp}-002").resolve(strict=True)
+    except FileNotFoundError:
+        logger.exception("pbench-receive-dir-prefix does not exist on the host")
+        sys.exit(1)
+    except Exception:
+        logger.exception(
+            "Exception occurred during setting up the upload directory on the host"
+        )
+        sys.exit(1)
+    else:
+        return upload_directory
+
+
+class ServerConfig(object):
+    CONFIG = get_config(os.environ.get("_PBENCH_SERVER_CONFIG"))
+
+    @property
+    def SERVER_CONFIG(self):
+        return self.CONFIG.conf["pbench-server"]
+
+    @property
+    def LOGGER(self):
+        return get_pbench_logger(__name__, self.CONFIG)
+
+    @property
+    def UPLOAD_DIRECTORY(self):
+        return get_prdp(self.SERVER_CONFIG, self.LOGGER)
+
+    @property
+    def PORT(self):
+        return self.SERVER_CONFIG.get("rest_port")
+
+    @property
+    def VERSION(self):
+        return self.SERVER_CONFIG.get("rest_version")
+
+    @property
+    def MAX_CONTENT_LENGTH(self):
+        return filesize_bytes(self.SERVER_CONFIG.get("rest_max_content_length"))
+
+    @property
+    def REST_URI(self):
+        return self.SERVER_CONFIG.get("rest_uri")
+
+    @property
+    def BIND_HOST(self):
+        return self.SERVER_CONFIG.get("bind_host")
+
+    @property
+    def WORKERS(self):
+        return self.SERVER_CONFIG.get("workers")
+
+
+class ProdServerConfig(ServerConfig):
+    """Production specific configuration"""
+
+    DEBUG = False
+    TESTING = False
+
+    @property
+    def ELASTICSEARCH_URI(self):
+        return f"http://{self.CONFIG.conf['elasticsearch'].get('host')}:{self.CONFIG.conf['elasticsearch'].get('port')}"
+
+    @property
+    def GRAPHQL_URI(self):
+        return f"http://{self.CONFIG.conf['graphql'].get('host')}:{self.CONFIG.conf['graphql'].get('port')}"
+
+    # Add prod specific db config
```

Comment:
For the others, it's arguable whether it makes more sense to pre-fetch and return local properties vs grabbing them from the composed config object. In these, however, I think that returning local properties instead of doing the string formatting on each call would make more sense.

Reply:
yeah I could define a property to get elasticsearch_host and elasticsearch_port separately but then where do we need these properties as separate things, each time we ask for es host and port its because we want to form a URI so I thought I would just return the URI because thats what most of the time we need

Reply:
No, what I meant is that it'd be (slightly) more efficient and a bit cleaner to construct this string in `__init__` and simply return the constructed property here rather than re-fetching the config options and rebuilding the template each time.